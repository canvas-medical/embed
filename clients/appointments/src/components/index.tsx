import { h } from 'preact'
import { useCallback, useEffect, useState } from 'preact/hooks'
import {
  AppointmentType,
  defaultAppointmentType,
  getAppointmentsList,
  Loader,
  putAppointment,
  Error,
  ErrorType,
  HandleErrorType,
  ProvidersType,
} from '@canvas-medical/embed-common'
import { IAppProps } from '../utils'

import { Ui } from './ui'

const noOp = () => {}

const defaultAppointment: AppointmentType = {
  id: '',
  code: '',
  description: '',
  display: '',
  locationId: '',
  providerId: '',
  start: '',
  end: '',
}

export const AppointmentsView = ({
  api,
  colors,
  callbacks,
  locationId,
  patientId,
  patientKey,
  providerIds,
  shadowRoot,
}: IAppProps) => {
  const [loading, setLoading] = useState(true)
  const [error, setError] = useState<ErrorType>()
  const [appointmentCancellation, setAppointmentCancellation] = useState({
    popoverOpen: false,
    appointment: defaultAppointment,
  })

  const [appointments, setAppointments] = useState<AppointmentType[]>([])
  const [providers, setProviders] = useState<ProvidersType[]>([])
  const [initialized, setInitialized] = useState<boolean>(false)
  const [loadStartTime] = useState(new Date())

  const handleCancel = (appointment: AppointmentType) => {
    setAppointmentCancellation({ popoverOpen: true, appointment })
  }

  const handleError: HandleErrorType = (error, msg) => {
    callbacks?.onError?.(error, msg)
    setError(msg)
    setLoading(false)
  }

  useEffect(() => {
    if (!loading && !initialized) {
      setInitialized(true)
      callbacks?.onLoad?.((new Date().getTime() - loadStartTime.getTime()))
    }
  }, [loading, initialized])

  const fetchAppointments = useCallback(() => {
    getAppointmentsList({
      setLoading,
      onError: handleError,
      setAppointments,
      setProviders,
      providerIds,
      api,
      patientId,
      patientKey,
    })
  }, [api, patientId, patientKey, providerIds])

  const afterCancel = () => {
    setAppointmentCancellation({
      popoverOpen: false,
      appointment: defaultAppointment,
    })
    fetchAppointments()
  }

  const onCancel = () => {
    try {
      putAppointment({
        onComplete: afterCancel,
        onError: handleError,
        setLoading,
        appointmentCoding: {
          code:
            appointmentCancellation?.appointment?.code ||
            defaultAppointmentType.code,
        },
        locationId:
          appointmentCancellation.appointment.locationId || locationId,
        timeSlot: {
          start: appointmentCancellation.appointment.start,
          end: appointmentCancellation.appointment.end,
          provider: {
            id: appointmentCancellation.appointment.providerId,
          },
        },
        patientId,
        patientKey,
        api,
        appointmentId: appointmentCancellation.appointment.id,
      })
    } catch (e) {
      handleError(e as Error, 'Error Cancelling Appointment')
    }
  }

  const onKeep = () => {
    setAppointmentCancellation({
      popoverOpen: false,
      appointment: defaultAppointment,
    })
  }

  useEffect(() => {
    fetchAppointments()
  }, [fetchAppointments])

  if (loading) {
    return <Loader colors={colors} />
  }

  if (error && error.length) {
    return <Error errorMessages={error} />
  }

  return (
    <Ui
      appointments={appointments}
      providers={providers}
      colors={colors}
      onCancel={onCancel}
      onKeep={onKeep}
      handleCancel={handleCancel}
      shadowRoot={shadowRoot}
      appointmentCancellation={appointmentCancellation}
    />
  )
}
