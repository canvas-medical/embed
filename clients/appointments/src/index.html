<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Appointments</title>
    <meta name="viewport" content="width=device-width,initial-scale=1" />
    <style>
      @import url('https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap');

      body {
        margin: 0;
      }
    </style>
  </head>
  <body>
    <div id="embed-root"></div>
    <script src="./appointments.js" type="text/javascript"></script>
    <script type="text/javascript">
      const callbacks = {
        onClick: (e, config) => {
          if (e && e.target) {
            // Prescribed method for getting the element clicked
            // const closestTarget = e.target.closest('[data-analytics-id]')
            // console.log(
            //   closestTarget.dataset,
            //   'e.target.dataset',
            //   config,
            //   'config'
            // )
          }
        },
        onChange: (e, config) => {
          if (e && e.target) {
            // const closestTarget = e.target.closest('[data-analytics-id]')
            // console.log(
            //   closestTarget.dataset,
            //   'e.target.dataset',
            //   config,
            //   'config'
            // )
          }
        },
        onError: (error, msg) => {
          // console.error('error', error, msg)
        },
        onLoad: () => {},
      }

      const config = {
        apiKey: 'API_KEY',
        patientId: 'PATIENT_ID',
        providerIds: ['PROVIDER_ID'],
      }

      const apiBaseUrl = 'http://localhost:3000'

      fetch(
        `${apiBaseUrl}/Auth?` +
          new URLSearchParams({
            key: config.apiKey,
            patient: config.patientId,
          })
      )
        .then(res => res.json())
        .then(data => {
          const patientKey = data.patient_key

          Appointments.init({
            api: apiBaseUrl,
            bailoutURL: 'https://canvasmedical.com',
            locationId: '1',
            patientId: config.patientId,
            patientKey: patientKey,
            rootId: 'embed-root',
            callbacks,
            // brandColor: '',
            // accentColor: '',
            // fontFamily: 'Arial, serif',
          })
        })
    </script>
  </body>
</html>
