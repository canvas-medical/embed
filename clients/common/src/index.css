*,
*:before,
*:after {
  box-sizing: border-box;
}

h1,
h2,
h3,
p,
ol,
ul,
fieldset,
legend {
  margin: 0;
  padding: 0;
  font-weight: normal;
}

button {
  cursor: pointer;
}

ul {
  list-style: none;
}
