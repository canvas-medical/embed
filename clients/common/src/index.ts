import css from './index.css'

export * from './assets'
export * from './components'
export * from './utils'
export { css }
