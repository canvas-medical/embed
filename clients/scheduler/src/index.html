<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Scheduler</title>
    <meta name="viewport" content="width=device-width,initial-scale=1" />
    <style>
      body {
        margin: 0;
        overflow: hidden;
      }
    </style>
    <link rel="preconnect" href="https://fonts.googleapis.com" />
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin />
    <link
      href="https://fonts.googleapis.com/css2?family=Roboto:wght@400;700&display=swap"
      rel="stylesheet"
    />
  </head>
  <body>
    <div id="embed-root"></div>
    <script src="./scheduler.js" type="text/javascript"></script>
    <script type="text/javascript">
      const callbacks = {
        onClick: (e, config) => {
          if (e && e.target) {
            // Prescribed method for getting the element clicked
            // const closestTarget = e.target.closest('[data-analytics-id]')
            // console.log(
            //   closestTarget.dataset,
            //   'e.target.dataset',
            //   config,
            //   'config'
            // )
          }
        },
        onChange: (e, config) => {
          if (e && e.target) {
            // const closestTarget = e.target.closest('[data-analytics-id]')
            // console.log(
            //   closestTarget.dataset,
            //   'e.target.dataset',
            //   config,
            //   'config'
            // )
          }
        },
        onError: (error, msg) => {
          // console.error('error', error, msg)
        },
        onLoad: () => {},
        onDateChange: config => {
          // console.log(config, 'config')
        },
      }

      const config = {
        apiKey: 'API_KEY',
        patientId: 'PATIENT_ID',
        providerIds: ['PROVIDER_ID'],
      }

      // This is the URL of the running proxy container
      const apiBaseUrl = 'http://localhost:3000'

      fetch(
        `${apiBaseUrl}/Auth?` +
          new URLSearchParams({
            key: config.apiKey,
            patient: config.patientId,
          })
      )
        .then(res => res.json())
        .then(data => {
          const patientKey = data.patient_key

          Scheduler.init({
            api: apiBaseUrl,
            appointmentCoding: {
              code: 'CODE',
              system: 'http://snomed.info/sct',
            },
            bailoutURL: 'https://canvasmedical.com',
            callbacks: callbacks,
            daysToFetch: '30',
            duration: '60',
            locationId: '1',
            patientId: config.patientId,
            patientKey: patientKey,
            providerIds: config.providerIds,
            reason: 'high fever, cough, leeches',
            returnURL: 'https://canvasmedical.com',
            rootId: 'embed-root',
            // brandColor: '',
            // accentColor: '',
            // customFont: 'Nunito',
          })
        })
        .catch(e => {
          console.error(e)
        })
    </script>
  </body>
</html>
