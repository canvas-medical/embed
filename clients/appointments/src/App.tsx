import { h } from 'preact'
import { AppContainer, Header } from '@canvas-medical/embed-common'
import { IAppProps } from './utils'
import { AppointmentsView } from './components'

export const App = ({
  api,
  bailoutURL,
  locationId,
  patientId,
  patientKey,
  providerIds,
  shadowRoot,
  colors,
  customFont,
}: IAppProps) => {
  return (
    <AppContainer customFont={customFont}>
      <Header
        bailoutURL={bailoutURL}
        colors={colors}
        title="Your Appointments"
      />
      <AppointmentsView
        api={api}
        bailoutURL={bailoutURL}
        colors={colors}
        locationId={locationId}
        patientId={patientId}
        patientKey={patientKey}
        providerIds={providerIds}
        shadowRoot={shadowRoot}
      />
    </AppContainer>
  )
}
