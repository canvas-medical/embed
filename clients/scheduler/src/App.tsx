import { h } from 'preact'
import { AppContainer, Body, Error, Header } from '@canvas/embed-common'
import { useAppContext } from './hooks'
import { Confirmation, DateSelect, TimeSlotSelect } from './components'

export const App = () => {
  const { bailoutURL, colors, screen, error } = useAppContext()
  return (
    <AppContainer>
      <Header
        bailoutURL={bailoutURL}
        colors={colors}
        title="Schedule an Appointment"
      />
      {error && error.length ? (
        <Error errorMessages={error} />
      ) : screen !== 'CONFIRM' ? (
        <Body>
          <DateSelect />
          <TimeSlotSelect />
        </Body>
      ) : (
        <Body>
          <Confirmation />
        </Body>
      )}
    </AppContainer>
  )
}
