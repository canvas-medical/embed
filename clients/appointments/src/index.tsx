import { h, render } from 'preact'
import { StyleSheetManager } from 'styled-components'
import { css, Error, generateColors } from '@canvas-medical/embed-common'
import { App } from './App'
import { IInitializerProps, IAppointmentProps, hasAllValues } from './utils'

export const Appointments = (props: IAppointmentProps) => {
  const {
    api,
    bailoutURL,
    locationId,
    patientId,
    patientKey,
    providerIds,
    brandColor,
    accentColor,
    shadowRoot,
    customFont,
  } = props
  const colors = generateColors(brandColor, accentColor)
  const allValuesProvided = hasAllValues(props)

  return (
    // Ignoring type mismatch error on target - ShadowRoot is an acceptable type
    // @ts-ignore
    <StyleSheetManager target={shadowRoot}>
      {!allValuesProvided.length ? (
        <App
          api={api}
          bailoutURL={bailoutURL}
          colors={colors}
          locationId={locationId}
          patientId={patientId}
          patientKey={patientKey}
          providerIds={providerIds}
          shadowRoot={shadowRoot}
          customFont={customFont}
        />
      ) : (
        <Error errorMessages={allValuesProvided} />
      )}
    </StyleSheetManager>
  )
}

export const init = ({
  api,
  bailoutURL,
  locationId,
  patientId,
  patientKey,
  providerIds,
  brandColor,
  accentColor,
  rootId,
  customFont,
}: IInitializerProps) => {
  const appRoot = document.querySelector(`#${rootId}`)

  if (!appRoot) {
    console.error('App root could not be found. Check your rootId')
    return null
  }

  appRoot.attachShadow({
    mode: 'open',
  })

  if (!appRoot.shadowRoot) {
    console.error('Shadow root could not be attached')
    return null
  }

  const styleTag = document.createElement('style')
  styleTag.innerHTML = css
  appRoot.shadowRoot.appendChild(styleTag)

  render(
    <Appointments
      api={api}
      bailoutURL={bailoutURL}
      locationId={locationId}
      patientId={patientId}
      patientKey={patientKey}
      providerIds={providerIds}
      brandColor={brandColor}
      accentColor={accentColor}
      shadowRoot={appRoot.shadowRoot}
      customFont={customFont}
    />,
    appRoot.shadowRoot
  )
}
