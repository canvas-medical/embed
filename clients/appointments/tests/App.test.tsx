import { generateColors } from '@canvas-medical/embed-common'
import { h } from 'preact'
import { render } from 'preact-render-to-string'
import { App } from '../src/App'

describe('App', () => {
  it('renders', () => {
    const colors = generateColors(null, null)

    expect(
      render(
        <App
          api="arbitrary.string"
          bailoutURL="arbitrary.com"
          locationId="1234"
          patientId="1234"
          patientKey="1234"
          shadowRoot="1234"
          colors={colors}
        />
      )
    ).toMatchSnapshot()
  })
})
