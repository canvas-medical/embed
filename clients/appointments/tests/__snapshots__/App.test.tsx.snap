// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `"<div class=\\"sc-gsDKAQ hcJErS\\"><div class=\\"sc-ksdxgE jAELwq\\"><div class=\\"sc-hBUSln eZVHDj\\"><a href=\\"arbitrary.com\\" class=\\"sc-egiyK gAFTBw\\"><svg height=\\"16\\" viewBox=\\"0 0 10 16\\" width=\\"10\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"m401.034779 24.5989021v10h10\\" fill=\\"none\\" stroke=\\"#fff\\" stroke-width=\\"2\\" transform=\\"matrix(.70710678 .70710678 -.70710678 .70710678 -257.145582 -299.440628)\\"></path></svg></a><div class=\\"sc-fotOHu bQqSAl\\"><h1 class=\\"sc-furwcr dnfnN\\">Your Appointments</h1></div></div></div><div class=\\"sc-bdvvtL biLwMd\\"><div class=\\"sc-hKwDye cQHjzB\\"><div class=\\"sc-fFeiMQ jjjCFF\\"></div></div></div></div>"`;
